src/mc-fle2-payload-iev.c

```diff
@@ -52,6 +55,230 @@ mc_FLE2IndexedEqualityEncryptedValueTokens_new (void)
+   BSON_ASSERT_PARAM (buf);
+
+   mc_reader_t reader;
+   mc_reader_init_from_buffer (&reader, buf, __FUNCTION__);
+
+   CHECK_AND_RETURN (mc_reader_read_u64 (&reader, &tokens->counter, status));
+
+   CHECK_AND_RETURN (
+      mc_reader_read_prfblock_buffer (&reader, &tokens->edc, status));
+
+   CHECK_AND_RETURN (
+      mc_reader_read_prfblock_buffer (&reader, &tokens->esc, status));
+
+   CHECK_AND_RETURN (
+      mc_reader_read_prfblock_buffer (&reader, &tokens->ecc, status));
+
+   return true;
+}
+
+static bool
+mc_fle2IndexedEncryptedValue_encrypt (
+   _mongocrypt_crypto_t *crypto,
+   const _mongocrypt_buffer_t *ClientEncryptedValue,
+   mc_ServerDataEncryptionLevel1Token_t *token,
+   mc_FLE2IndexedEqualityEncryptedValueTokens *index_tokens,
+   _mongocrypt_buffer_t *out,
+   mongocrypt_status_t *status);
+
+bool
+safe_uint32_t_sum (const uint32_t a,
+                   const uint32_t b,
+                   uint32_t *out,
+                   mongocrypt_status_t *status)
+{
+   if (a > UINT32_MAX - b) {
+      CLIENT_ERR ("safe_uint32_t_sum overflow, %" PRIu32 ", %" PRIu32, a, b);
+      return false;
+   }
+   *out = a + b;
+   return true;
+}
+
+bool
+mc_FLE2IndexedEncryptedValue_write (
+   _mongocrypt_crypto_t *crypto,
+   const bson_type_t original_bson_type,
+   const _mongocrypt_buffer_t *S_KeyId,
+   const _mongocrypt_buffer_t *ClientEncryptedValue,
+   mc_ServerDataEncryptionLevel1Token_t *token,
+   mc_FLE2IndexedEqualityEncryptedValueTokens *index_tokens,
+   _mongocrypt_buffer_t *buf,
+   mongocrypt_status_t *status)
+{
+   BSON_ASSERT_PARAM (crypto);
+   BSON_ASSERT_PARAM (index_tokens);
+   BSON_ASSERT_PARAM (S_KeyId);
+   BSON_ASSERT_PARAM (ClientEncryptedValue);
+   BSON_ASSERT_PARAM (token);
+   BSON_ASSERT_PARAM (index_tokens);
+   BSON_ASSERT_PARAM (buf);
+
+   if (ClientEncryptedValue->len == 0) {
+      CLIENT_ERR (
+         "mc_FLE2IndexedEncryptedValue_write iev must have an encrypted value");
+      return false;
+   }
+
+   if (S_KeyId->len == 0) {
+      CLIENT_ERR (
+         "mc_FLE2IndexedEncryptedValue_write iev SKeyId must have value");
+      return false;
+   }
+
+   _mongocrypt_buffer_t encryption_out;
+   _mongocrypt_buffer_init (&encryption_out);
+
+   if (!mc_fle2IndexedEncryptedValue_encrypt (crypto,
+                                              ClientEncryptedValue,
+                                              token,
+                                              index_tokens,
+                                              &encryption_out,
+                                              status)) {
+      return false;
+   }
+
+   uint32_t expected_plaintext_size = 0;
+   CHECK_AND_RETURN (safe_uint32_t_sum (
+      ClientEncryptedValue->len,
+      (uint32_t) (sizeof (uint64_t) * 2 + sizeof (uint32_t) * 3),
+      &expected_plaintext_size,
+      status));
+
+   uint32_t expected_cipher_size = _mongocrypt_fle2_calculate_ciphertext_len (
+      expected_plaintext_size, status);
```

Comment:
Missing error handling on potential failure of `_mongocrypt_fle2_calculate_ciphertext_len`. If guaranteed to not fail, suggest adding a comment or `BSON_ASSERT()`.